src/storage/lock_manager/lock_waiting_queue.rs

```diff
@@ -0,0 +1,950 @@
+
+pub struct LockWaitQueueInner {
+    queue_map: dashmap::DashMap<Key, KeyLockWaitState>,
+    id_allocated: AtomicU64,
+}
+
+#[derive(Clone)]
+pub struct LockWaitQueues<L: LockManager> {
+    inner: Arc<LockWaitQueueInner>,
+    #[allow(dead_code)]
+    lock_mgr: L,
+}
+
+impl<L: LockManager> LockWaitQueues<L> {
+    pub fn new(lock_mgr: L) -> Self {
+        Self {
+            inner: Arc::new(LockWaitQueueInner {
+                queue_map: dashmap::DashMap::new(),
+                id_allocated: AtomicU64::new(1),
+            }),
+            lock_mgr,
+        }
+    }
+
+    /// Enqueues a lock wait entry. The key is indicated by the `key` field of
+    /// the `lock_wait_entry`. The caller also needs to provide the
+    /// information of the current-holding lock.
+    pub fn push_lock_wait(
+        &self,
+        mut lock_wait_entry: Box<LockWaitEntry>,
+        current_lock: kvrpcpb::LockInfo,
+    ) {
+        let mut entry = self
+            .inner
+            .queue_map
+            .entry(lock_wait_entry.key.clone())
+            .or_insert_with(|| KeyLockWaitState::new());
+        entry.current_lock = current_lock;
+
+        if lock_wait_entry.current_legacy_wake_up_cnt.is_none() {
+            lock_wait_entry.current_legacy_wake_up_cnt = Some(entry.value().legacy_wake_up_cnt);
+        }
+        entry.value_mut().queue.push(lock_wait_entry);
+    }
+
+    /// Dequeues the head of the lock waiting queue of the specified key,
+    /// assuming the popped entry will be woken up.
+    ///
+    /// If it's waking up a legacy request and the queue is not empty, a future
+    /// will be returned and the caller will be responsible for executing it.
+    /// The future waits until `wake_up_delay_duration` is elapsed since the
+    /// most recent waking-up, and then wakes up all lock waiting entries that
+    /// exists at the time when the latest waking-up happens. The future
+    /// will return a `LockWaitEntry` if a resumable entry is popped out
+    /// from the queue while executing, and in this case the caller will be
+    /// responsible to wake it up.
+    pub fn pop_for_waking_up(
+        &self,
+        key: &Key,
+        conflicting_start_ts: TimeStamp,
+        conflicting_commit_ts: TimeStamp,
+        wake_up_delay_duration_ms: u64,
+    ) -> Option<(Box<LockWaitEntry>, Option<DelayedNotifyAllFuture>)> {
+        self.pop_for_waking_up_impl(
+            key,
+            conflicting_start_ts,
+            conflicting_commit_ts,
+            Some(wake_up_delay_duration_ms),
+        )
+    }
+
+    fn pop_for_waking_up_impl(
+        &self,
+        key: &Key,
+        conflicting_start_ts: TimeStamp,
+        conflicting_commit_ts: TimeStamp,
+        wake_up_delay_duration_ms: Option<u64>,
+    ) -> Option<(Box<LockWaitEntry>, Option<DelayedNotifyAllFuture>)> {
+        let mut result = None;
+
+        // We don't want other threads insert any more entries between finding the
```

Comment:
I don't have comment on `remove_if_mut` here. But I'm curious about when there could be concurrent insert here. The latch should prevent concurrent locking requests?

Reply:
Yes, for `pop_for_waking_up` function in current implementation.  I think doing this makes it less coupling with Scheduler's logic that makes use of the lock waiting queue.

By the way I think there might be some more possible usages of it in the future. For example, when a new pessimistic lock request finds that the key is locked by checking the lock waiting queue, it can be put to the queue without entering scheduler and reading the data. In this case, an entry might be inserted without holding the latch. (This optimization is currently just an incomplete idea).

By the way*2, `delayed_notify_all` also uses `remove_if_mut` and that function won't be executed while holding latch.